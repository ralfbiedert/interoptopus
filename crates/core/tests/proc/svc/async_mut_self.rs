use interoptopus::{
    ffi,
    pattern::asynk::{Async, AsyncRuntime},
};
use std::future::Future;

#[ffi]
enum Error {
    Something,
}

#[ffi(service)]
struct Service;

#[ffi]
impl Service {
    pub async fn new(_: Async<Self>) -> ffi::Result<u8, Error> {
        ffi::Ok(0)
    }

    pub fn bad(&mut self) {}
}

impl AsyncRuntime for Service {
    type T = ();

    fn spawn<Fn, F>(&self, _f: Fn)
    where
        Fn: FnOnce(()) -> F,
        F: Future<Output = ()> + Send + 'static,
    {
    }
}

fn main() {}
