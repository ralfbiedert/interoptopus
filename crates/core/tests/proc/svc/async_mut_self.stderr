error: Async services cannot have methods with &mut self. Use &self instead.
  --> tests/proc/svc/async_mut_self.rs:21:16
   |
21 |     pub fn bad(&mut self) {}
   |                ^^^^^^^^^
